File: src/main/kotlin/travelator/money/Money.java
package travelator.money;

import java.math.BigDecimal;
import java.util.Currency;
import java.util.Objects;

public class Money {
  private final BigDecimal amount;
  private final Currency currency;

  private Money(BigDecimal amount, Currency currency) {
    this.amount = amount;
    this.currency = currency;
  }

  public static Money of(BigDecimal amount, Currency currency) {
    return new Money(
        amount.setScale(currency.getDefaultFractionDigits()),
        currency
    );
  }

  public BigDecimal getAmount() {
    return amount;
  }

  public Currency getCurrency() {
    return currency;
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) {
      return true;
    }
    if (!(o instanceof Money money)) {
      return false;
    }
    return amount.equals(money.getAmount()) && currency.equals(money.getCurrency());
  }

  @Override
  public int hashCode() {
    return Objects.hash(getAmount(), getCurrency());
  }

  @Override
  public String toString() {
    return amount.toString() + " " + currency.getCurrencyCode();
  }

  public Money add(Money that) {
    if (!this.currency.equals(that.currency)) {
      throw new IllegalArgumentException(
          "can not add Money values of different currencies"
      );
    }
    return new Money(that.amount.add(that.amount), that.currency);
  }

}

File: src/main/kotlin/travelator/money/Money.kt
package travelator.money

import java.math.BigDecimal
import java.util.Currency
import java.util.Objects

/**
 * 자바와 비교해서 가독성이 좋다고 평가할 수 없다.
 * - 정적팩터리함수로만 생성 가능하게 열어두고 BigDecimal의 Scale(정밀도)에 제한을 두었다. 하지만 이 클래스를 데이터 클래스로
 * 변환하면 자동생성되는 copy 함수 때문에 불변성이 깨질 수 있다.
 */

class Money
private constructor(
    val amount: BigDecimal,
    val currency: Currency,
){

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        val money = other as Money
        return amount == money.amount && currency == money.currency
    }

    override fun hashCode(): Int {
        return Objects.hash(amount, currency)
    }

    override fun toString(): String {
        return amount.toString() + " " + currency.currencyCode
    }

    fun add(that: Money): Money {
        require(currency == that.currency) {
            "cannot add Money values of different currencies"
        }
        return Money(amount.add(that.amount), currency)
    }

    companion object {
        @JvmStatic
        fun of(amount: BigDecimal, currency: Currency): Money {
            return Money(
                amount.setScale(currency.defaultFractionDigits),
                currency,
            )
        }
    }
}